### PR TITLE
[IMP] survey: graph conditions change for MCQs with one answer

### PR DESCRIPTION
**Pull request by vivp-odoo**
This PR improves the behavior of MCQs with one answer. Previously, when
there were multiple correct answers and any one of them was selected,
a "partial" graph was displayed instead of a "correct" graph. With this PR,
the behavior has been modified as follows:

- If the selected option is correct and has any score, the "correct" graph will
be shown.
- If the selected option is incorrect but scores are above zero, the "partial"
graph will be shown.
- If the score is zero or less and the option is not correct, the "incorrect"
graph will be shown.

**Task**- 3252605.

### CODE REVIEW

File: addons/survey/models/survey_user_input.py [hunk 1]
```diff
@@ -444,8 +456,27 @@ def _prepare_statistics(self):
...
+                incorrect_user_input_line = line.suggested_answer_id
+
+        if correct_user_input_line in question_correct_suggested_answers:
+            return 'correct'
+        elif incorrect_user_input_line in question_scored_incorrect_answers:
+            return 'partial'
+        elif not correct_user_input_line and incorrect_user_input_line:
+            return 'incorrect'
+        else:
+            return 'skipped'
```
Comment (Odoonan) [inline addons/survey/models/survey_user_input.py:477]: Should it not, according to spec, be:


```Python3
  if correct_user_input_answer in question_correct_suggested_answers:
      return 'correct'
  elif incorrect_user_input_answer in question_incorrect_scored_answers:
      return 'partial'
  elif incorrect_user_input_answer:
      return 'incorrect'
  else:
      return 'skipped'
```

File: addons/survey/models/survey_user_input.py [hunk 2]
```diff
@@ -401,14 +401,26 @@ def _prepare_statistics(self):
         scored_questions = self.mapped('predefined_question_ids').filtered(lambda question: question.is_scored_question)
```
[the remark this answers is not in the document]
Comment (Odoonan) [reply addons/survey/models/survey_user_input.py:401]: Also, maybe to ask to FRGI / LNA, when using 'review your answers' we get for instance this : https://tinyurl.com/2q9p5ey2 where you can see the background is green on the third choice (because score is greater than 0, but the answer is not correct.) Is this desired? (could be a task / fix in master or before in separate task) (other ongoing tasks may already change that)

**Review by Odoonan — COMMENTED** [review]
Hello @vivp-odoo ,

Here is Validation, good job overall, I've mainly added suggestions to reduce the size of the PR and a few general comments.

Cheers!

File: addons/survey/tests/test_survey.py [hunk 3]
```diff
@@ -144,6 +144,110 @@ def test_partial_scores_simple_choice(self):
...
+        test_survey = self.env['survey.survey'].create({
+            'title': 'Test This Survey',
+            'scoring_type': 'scoring_with_answers',
+            'scoring_success_min': 80.0,
+        })
+
+        [a_01, a_02, a_03, a_04] = self.env['survey.question.answer'].create([{
+            'value': 'In Europe',
+            'answer_score': 0.0,
+            'is_correct' : False
```
[the remark this answers is not in the document]
Comment (Odoonan) [reply addons/survey/tests/test_survey.py:156]: @vivp-odoo, can you remove the space before the `:` ? thanks :p

File: addons/survey/views/survey_templates.xml [hunk 4]
```diff
@@ -432,7 +432,7 @@
                 <t t-foreach='question.suggested_answer_ids' t-as='label'>
                     <t t-set="item_idx" t-value="label_index"/>
                     <t t-set="answer_selected" t-value="answer_line and answer_line.suggested_answer_id.id == label.id"/>
-                    <t t-set="is_correct" t-value="label.answer_score > 0.0"/>
+                    <t t-set="is_correct" t-value="label.is_correct"/>
```
Comment (Odoonan) [inline addons/survey/views/survey_templates.xml:435]: Can you please put it in a separate commit? And do the change in question_multiple_choice template as well (see below) ?

Comment (Odoonan) [inline addons/survey/views/survey_templates.xml:435]: Commit guidelines: use a verb.
Suggestion: [FIX] survey: use bg-success only if answer is_correct

File: addons/survey/models/survey_user_input.py [hunk 5]
```diff
@@ -401,14 +401,19 @@ def _prepare_statistics(self):
...
```
Comment (Odoonan) [inline addons/survey/models/survey_user_input.py:402]: Commit message guidelines.
Suggestion: [FIX] survey: change simple answer MCQ graph values

File: addons/survey/views/survey_templates.xml (hunk 4 above)
Comment (Odoonan) [reply addons/survey/views/survey_templates.xml:435]: You can also add task-... in the commit message.
In the message, you can also add that the change concerns MCQs.

**Review by Odoonan — COMMENTED** [review]
Hello @vivp-odoo ,

We are almost there :) Can you do these small commit message improvements?

Cheers

File: addons/survey/views/survey_templates.xml (hunk 4 above)
Comment (vivp-odoo) [reply addons/survey/views/survey_templates.xml:435]: @Odoonan Done

File: addons/survey/tests/test_survey.py [hunk 6]
```diff
@@ -144,6 +144,66 @@ def test_partial_scores_simple_choice(self):
...
+
+        user_input = self.env['survey.user_input'].create({'survey_id': test_survey.id})
+        user_input_line = self.env['survey.user_input.line'].create({
+            'user_input_id': user_input.id,
+            'question_id': q_01.id,
+            'answer_type': 'suggestion',
+            'suggested_answer_id': a_01.id
+        })
+
+        def assertOnlyAnswerInStatsIs(status, stats):
```
Comment (awa-odoo) [inline addons/survey/tests/test_survey.py:186]: We usually don't use inline methods, can you put it as a regular tooling method at the end of this test instead?

Comment (awa-odoo) [inline addons/survey/tests/test_survey.py:186]: Sugg:

`assertAnswerStatus(questions_statistics, expected_answer_status)`

File: addons/survey/tests/test_survey.py [hunk 7]
```diff
@@ -144,6 +144,66 @@ def test_partial_scores_simple_choice(self):
...
+            ''' Assert counts for 'Correct', 'Partially', 'Incorrect', 'Unanswered' are 0, and 1 for status one '''
```
Comment (awa-odoo) [inline addons/survey/tests/test_survey.py:187]: "... and 1 for our expected status"

File: addons/survey/tests/test_survey.py [hunk 8]
```diff
@@ -144,6 +144,66 @@ def test_partial_scores_simple_choice(self):
...
+            self.assertTrue(all([
+                total['count'] == int(total['text'] == status) for total in stats['totals']
+            ]), f'The only answer should be {status} answer for this user input')
```
Comment (awa-odoo) [inline addons/survey/tests/test_survey.py:190]: So we are casting a boolean to int to get 0 or 1 ? A little magic.

Would be a bit easier to read with:

```py
for status, count in [(total['text'], total['count']) for total in stats['totals']]:
    self.assertEqual(count, 1 if status == expected_status else 0)
```

File: addons/survey/models/survey_user_input.py [hunk 9]
```diff
@@ -457,6 +462,18 @@ def _choice_question_answer_result(self, user_input_lines, question_correct_sugg
...
+        correct_user_input_answer = user_input_lines.filtered(lambda line: line.answer_is_correct and not line.skipped).mapped('suggested_answer_id')
+        incorrect_user_input_answer = user_input_lines.filtered(lambda line: not line.answer_is_correct and not line.skipped).mapped('suggested_answer_id')
+        if correct_user_input_answer in question_correct_suggested_answers:
+            return 'correct'
+        elif incorrect_user_input_answer in question_incorrect_scored_answers:
+            return 'partial'
+        elif incorrect_user_input_answer:
+            return 'incorrect'
+        else:
+            return 'skipped'
```
Comment (awa-odoo) [inline addons/survey/models/survey_user_input.py:475]: I mean we only have 1 answer for those right? So this can be simplified to:

```py
# only 1 answer here
user_input_line = user_input_lines[0]
user_answer = user_input_line.suggested_answer_id if not user_input_line.skipped else self.env['survey.question.answer']
if user_answer in question_correct_suggested_answers:
    return 'correct'
elif user_answer in question_incorrect_scored_answers:
    return 'partial'
elif user_answer:
    return 'incorrect'
else:
    return 'skipped'
```

**Review by awa-odoo — COMMENTED** [review]
Hello @vivp-odoo 

Left a few remarks to try to polish.

However I don't think we should do this in 15.2, but rather on master.
But let's discuss this on the task pad.

Thanks & Cheers!

**Comment by vivp-odoo** [comment]
Hello @awa-odoo 
I have made the changes as per your suggestion.
Will you please have a look at it?

Thanks!!

**Comment by vivp-odoo** [comment]
Hello @awa-odoo 
I have re-targeted this PR to the "master" branch as it is considered an improvement. Regarding the issue raised by @Odoonan , which was my second commit earlier, it is recognized as a fix that needs to be addressed from version 14.0 up to saas-15.3. However, for the above versions, there is no need for a fix as it has already been resolved. Here is the link to the PR for the second commit: https://github.com/odoo/odoo/pull/123878.

Thanks!!

File: addons/survey/tests/test_survey.py [hunk 10]
```diff
@@ -271,6 +271,66 @@ def test_partial_scores_simple_choice(self):
...
+
+        user_input = self.env['survey.user_input'].create({'survey_id': test_survey.id})
+        user_input_line = self.env['survey.user_input.line'].create({
+            'user_input_id': user_input.id,
+            'question_id': q_01.id,
+            'answer_type': 'suggestion',
+            'suggested_answer_id': a_01.id
+        })
+
+        statistics = user_input._prepare_statistics()[user_input]
```
Comment (awa-odoo) [inline addons/survey/tests/test_survey.py:313]: Would add a small comment above all of the checks:

```
# this answer is incorrect with no score: should be considered as incorrect
...
# this answer is correct with a positive score (even if not the maximum): should be considered as correct
...
# this answer is correct with the best score: should be considered as correct
...
# this answer is incorrect but has a score: should be considered as "partially"
```

**Review by awa-odoo — COMMENTED** [review]
Hello @vivp-odoo 

Left 2 last small remarks and then we should be good to go.

Thanks & cheers!

**Comment by awa-odoo** [comment]
@robodoo r+
